compiler/rustc_parse/src/parser/pat.rs

```diff
@@ -395,7 +391,7 @@ impl<'a> Parser<'a> {
             } else {
                 PatKind::Lit(const_expr)
             }
-        } else if self.can_be_ident_pat() {
+        } else if self.can_be_ident_pat() || self.is_lit_bad_ident().is_some() {
```

Comment:
This code adds (not really adds, it was already here with the early return previously which was also a mistake but now is the second best moment to fix it) a parser regression.

```rust
macro_rules! pat {
    ($p:pat) => {};
}

fn main() {
    pat!(3meow);
}
```
This code should compile (as literals are valid patterns) but doesn't. For reference, replacing `:pat` with `:expr` makes it compile. It also compiles on stable, where the early return above hasn't landed yet.

This can be fixed by adding a `self.may_revover() &&` before this check. I still don't exactly like it, but it should fix the regression. I would prefer it if this wasn't an eager recovery but instead only started to influence behavior once there truly was an error, but I can accept it if you add the `may_recover()` and don't want to refactor it further.

For the future, always remember to think about how such changes can influence parser behavior and make sure to gate it behind a `may_recover()` if there the parser isn't in an error state yet.

Reply:
Right, so I was the author of that suggestion. Bear in mind that I don't have years on experience working on rustc but I don't think this is a regression.

If you compile the following code:
```rust
macro_rules! pat {
    ($p:pat) => {
        let $p = 5;
    };
}

fn main() {
    pat!(3meow);
}
```
(the same code but *using* the `$p` metavariable)

Then it emits an error. This is the case since `1.0.0`, the *only* thing this suggestion does is pick up on *always* invalid code and provide a better error message.

Secondly, AFAIK `Parser::may_recover` isn't "correct" here. Technically speaking, it is only for eager token recovery which neither the suggestion PR nor this PR introduce that. (eager recovery meaning consuming *multiple* tokens that might be valid?)

> if there the parser isn't in an error state yet.

We are in an error state though, a numeric literal with an invalid suffix is *always* invalid.


Maybe I'm completely wrong (\*cough\* like #107813 \*cough\*) though.

Reply:
> Secondly, AFAIK Parser::may_recover isn't "correct" here. Technically speaking, it is only for eager token recovery which neither the suggestion PR nor this PR introduce that.

You are absolutely right. Redirecting this to am early error like this is always wrong.

> a numeric literal with an invalid suffix is always invalid.

This is not quite correct. A literal with an invalid syntax is _semantically_ invalid. This means that it's not allowed in post expansion Rust code, as shown in your example which correctly errors. But an invalid suffix is _syntactically_ valid, so we can't error out because macros might delete it like in my example.
I am not really sure about the best way to get the nice diagnostic without introducing regressions. Maybe finding out where the example normally errors and then trying to add something there?

But don't worry about making mistakes here, introducing parser regressions like this happens to many others as well, it's hard to spot unless you're already aware of the potential problem^^

Reply:
You mean literals instead of expressions? That does sound a little tricky, adding a parameter would fix it but is also probably a little much.
So I guess the alternatives here are also having the diagnostic in other places where literals are allowed or not having it at all. I don't want to just put broken code behind a FIXME as these usually don't get fixed in quite a while.

Maybe you have some other ideas or you could try out adding the parameter if possible and see how bad it is.

Reply:
You're right, literals not expressions.

There wouldn't be broken code, just that the nice diagnostic would be not as nice in macros.

Reply:
The problem is that `may_recover` isn't set when parsing input to attribute proc macros (because this input _has_ to parse normally so eager recovery is fine there). But the early error here is not fine, as the macro might expand to nothing. So the "correct" heuristic would be "is not a macro input in any way" but that would rule out any attributes on the item, which is way too strict here.
These early errors also generally lead to more confusing code in the parser, so I would really prefer to avoid them (and with the current parser machinery we have, we must avoid them).

Reply:
As a sidenote, I would appreciate a PR renaming `may_recover` to something more descriptive like `may_recover_lookahead` or something like that to be more clear about what it does and doesn't affect.

Reply:
Hmm, so what do we do from here?

Reply:
I'll leave it up to you whether you would prefer removing the diagnostic or whether you'd be fine with changing it so that it also shows the note inside expressions. I don't think it would hurt, so I'd be fine with either.

In the meantime, you could also split out the first and last commit into a separate PR if you like, I would approve that.

Reply:
Are you sure `may_recover` wouldn't work here? It'll be a bit overreaching (all use in macros won't have the improved diagnostic) but it'll fix the regression.

Reply:
I think so, but I'm not entirely sure, I would need to check.
But actually, I just changed my mind about this PR. While this doesn't fix the regression, it doesn't introduce a new one either. We should merge this and I'll open an issue about the regression (which you can claim if you want, but of course don't have to).
We can continue this discussion on the issue.

Reply:
And `may_recover` is probably better than nothing, so let's do that anyways. It's not like this is a very critical issue anyways.